### ISSUE TITLE
TensorFlow compile failure with error tensorflow/tensorflow/compiler/xla/service/gpu/BUILD:649:1: C++ compilation of rule '//tensorflow/compiler/xla/service/gpu:infeed_manager' failed (Exit 1)

### ISSUE THREAD

**Issue by kami93** [tensorflow/tensorflow#22269]
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1. It must be a bug, a feature request, or a significant problem with documentation (for small docs fixes please send a PR instead).
2. The form below must be filled out.
3. It shouldn't be a TensorBoard issue. Those go [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 18.04.1 (GNU/Linux 4.15.0-34-generic x86_64)
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: No
- **TensorFlow installed from (source or binary)**: Source
- **TensorFlow version (use command below)**: 1.10.1
- **Python version**: Anaconda 5.2.0 (Python 3.6.5)
- **Bazel version (if compiling from source)**: 0.16.1
- **GCC/Compiler version (if compiling from source)**:7.3.0
- **CUDA/cuDNN version** : 9.2 / 7.2.1
- **GPU model and memory**: NVIDIA 1080Ti x 4 with driver 396.54 installed from ubuntu ppa.
- **Exact command to reproduce**: bazel build --config=opt --config=cuda //tensorflow/tools/pip_package:build_pip_package

### Describe the problem
Hi. I have a problem while compiling TensorFlow 1.10. I have always compiled pip package for my computer since back in TF versions like 0.12. But from TF 1.10 I started getting compile error related to XLA (I enabled XLA in ./configure before compiling). I did not get any XLA related error while compiling until TF version 1.9 but the error stated occurring from TF 1.10.

The error lines are the following:
TensorFlow fails to compile with error tensorflow/tensorflow/compiler/xla/service/gpu/BUILD:649:1: C++ compilation of rule '//tensorflow/compiler/xla/service/gpu:infeed_manager' failed (Exit 1)

**Comment by kami93** [comment]
> The error lines you copied are simply reports by bazel on roughly where the compilation failure happened.
> Here are a few things before we can help you:
> 
> * Try a more recent version of the source. Bazel is still pre 1.0, which means it has a lot of backwards incompatible changes. With your bazel version (0.16.1) it is possible the combination will just not work.
> * Share the full bazel output using pastebin, or a similar service.

Thank you for your response.
In accordance with your instructions, I have updated the Bazel up-to-date version (0.17.2). Unfortunately, the same error is still occuring.

Here is the full (as much as I could copy from bash) output.
[https://pastebin.com/j2PXPbBa](https://pastebin.com/j2PXPbBa)

Also, here is the full configure that I used for the compilation.
[https://pastebin.com/syJmPGPb](https://pastebin.com/syJmPGPb)

Have a nice day.

**Comment by gunan** [comment]
OK, this gives me much more information, thanks!
Here is the compilation error:
```
ERROR: /home/simon/Desktop/tensorflow/tensorflow/compiler/xla/service/gpu/BUILD:649:1: C++ compilation of rule '//tensorflow/compiler/xla/service/gpu:infeed_manager' failed (Exit 1)
In file included from ./tensorflow/compiler/xla/service/gpu/infeed_manager.h:23:0,
                 from tensorflow/compiler/xla/service/gpu/infeed_manager.cc:16:
./tensorflow/compiler/xla/service/gpu/xfeed_queue.h:68:37: error: 'std::function' has not been declared
```

Looking at that last part of the error, I found this:
https://www.bountysource.com/issues/49075870-compilation-error-loader-resourceloader-h-101-100-error-std-function-has-not-been-declared-loader-resourceloader-h-101-108-error-expected-or-before-token

Looks like this can be something in latest gcc, @jlebar @Artem-B to weigh in on that.

One thing you can try.
In our configuration script, it asks you if you like to use "clang" to compile:
`Do you want to use clang as CUDA compiler?`
Could you pick "clang" to compile, and when it asks, you will need to tell it to download and use latest clang.

**Comment by jlebar** [comment]
```
$ git log --patch master tensorflow/compiler/xla/service/gpu/xfeed_queue.h
commit fb8d1ca4eaefe58d42c27b6fc676f64f137f4675
Author: Ray Kim <msca8h@naver.com>
Date:   Sat Jul 21 21:42:15 2018 +0900

    fixed build error on gcc-7

diff --git a/tensorflow/compiler/xla/service/gpu/xfeed_queue.h b/tensorflow/compiler/xla/service/gpu/xfeed_queue.h
index 737c7eb025..dd46ff433b 100644
--- a/tensorflow/compiler/xla/service/gpu/xfeed_queue.h
+++ b/tensorflow/compiler/xla/service/gpu/xfeed_queue.h
@@ -17,6 +17,7 @@ limitations under the License.
 #define TENSORFLOW_COMPILER_XLA_SERVICE_GPU_XFEED_QUEUE_H_
 
 #include <deque>
+#include <functional>
 #include <vector>
 
 #include "tensorflow/core/platform/mutex.h"
```

Looks promising.  And looks like this change is not in v1.10.1:

```
$ git log --pretty=oneline v1.10.1 tensorflow/compiler/xla/service/gpu/xfeed_queue.h
eccc1d4c102b1b3b03b98dbc362f799cb540a1da [XLA:GPU] Unify infeed and outfeed queue implementations
```

**Comment by dksb** [comment]
Is this still an issue?

**Comment by jlebar** [comment]
I don't think it's an issue at head.

If you'd like to request a backport to  TF 1.10, I guess that's a separate bug?

**Comment by kami93** [comment]
@gunan 
I tried using clang for the CUDA compiler however the compiling process does not even start with the clang option. I guess this is a separate issue.

@jlebar @dksb 
The issue is resolved if I cherrypick fb8d1ca4eaefe58d42c27b6fc676f64f137f4675. Or just checkout v1.11.0. Thank you all for the kind helps.
